File: distributed-training/kubernetes/docker/worker/worker.js
const tf = require('@tensorflow/tfjs-node');
const amqp = require("amqplib");
const http = require("http");
const fs = require('fs');
var os = require("os");

const CACHE_STORAGE = "./_runtime/cache/"
const MODEL_STORAGE = "shared/models/"

const queueUrl = "amqp://rabbitmq-service:5672";
const inputQueue = process.env.JOB_NAME + "-INPUT";
const outputQueuePrefix = process.env.JOB_NAME + "-OUTPUT";

function guidGenerator() {
  var S4 = function () {
    return (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
  };
  return (S4() + S4() + "-" + S4() + "-" + S4() + "-" + S4() + "-" + S4() + S4() + S4());
}

function optimizerBuilderFunction(optimizer, learningRate) {
  var res;
  switch (optimizer) {
    case "sgd":
      res = tf.train.sgd(learningRate)
      break;
    case "adagrad":
      res = tf.train.adagrad(learningRate)
      break;
    case "adadelta":
      res = tf.train.adadelta(learningRate)
      break;
    case "adam":
      res = tf.train.adam(learningRate)
      break;
    case "adamax":
      res = tf.train.adamax(learningRate)
      break;
    case "rmsprop":
      res = tf.train.rmsprop(learningRate)
      break;
  }
  return res;
}

class DataSet {
  /**
   * 
   * @param {*} host 
   * @param {*} path 
   * @param {*} port 
   * @param {*} options { first: val, last: val }; first - Take only first portion of the data set. Acceptable value range 0 - 100; last - Take only last portion of the data set. Acceptable value range 0 - 100
   */
  constructor(host, path, port, cache_id, cache_batch_size, options) {
    this.host = host;
    this.path = path;
    this.port = port;
    this.cache_id = cache_id;
    this.cache_batch_size = cache_batch_size;
    this.index = 0;
    var defaultOptions = { first: 0, last: 0, batch_size: 32 };
    this.options = Object.assign(defaultOptions, options);
    if (parseInt(this.cache_batch_size) <= parseInt(this.options.batch_size))
      throw Error("cache_batch_size should always be bigger than options.batch_size");
  }

  async getNextBatchFunction() {
    var itemsCnt = 0;
    var lastLoadedCacheBatch = null;
    var lastLoadedCacheBatchIndex = null;
    if (this.cache_id) {
      var cacheFileDir = `${CACHE_STORAGE}${this.cache_id}/`;
      if (!fs.existsSync(cacheFileDir)) {
        var tmpFolderGuid = guidGenerator();
        cacheFileDir = `${CACHE_STORAGE}${tmpFolderGuid + "_"}${this.cache_id}/`;
        fs.mkdirSync(cacheFileDir, { recursive: true });
        var batch_index = 0;
        var asyncRequest = () => {
          var options = {
            host: this.host,
            port: this.port,
            path: `${this.path}?index=${batch_index}&cache_batch_size=${this.cache_batch_size}`,
            method: 'GET'
          };
          return new Promise(function (resolve, reject) {
            var req = http.get(options, function (res) {
              var bodyChunks = [];
              res.on('data', function (chunk) {
                bodyChunks.push(chunk);
              }).on('end', function () {
                var body = Buffer.concat(bodyChunks);
                resolve(JSON.parse(body));
              })
            });

            req.on('error', function (e) {
              console.log('ERROR: ' + e.message);
              reject(e);
            });
          });
        }
        var res;
        itemsCnt = 0;
        do {
          var res = await asyncRequest(batch_index);
          fs.writeFileSync(`${cacheFileDir}${batch_index}.json`, res.value);
          itemsCnt += JSON.parse(res.value).xs.length;
          batch_index++;
          if (res.done) {
            var firstBatch = JSON.parse(fs.readFileSync(`${cacheFileDir}0.json`));
            firstBatch.itemsCnt = itemsCnt;
            fs.writeFileSync(`${cacheFileDir}0.json`, JSON.stringify(firstBatch))
          }
        } while (!res.done);
        console.log('completed downloading. renaming folder');
        fs.renameSync(cacheFileDir, `${CACHE_STORAGE}${this.cache_id}/`);
        console.log('completed renaming folder');
        cacheFileDir = `${CACHE_STORAGE}${this.cache_id}/`;
      }

      this.options.first = parseFloat(this.options.first);
      this.options.last = parseFloat(this.options.last);

      if (!itemsCnt) {
        var entierCacheBatch = JSON.parse(fs.readFileSync(`${cacheFileDir}0.json`));
        itemsCnt = entierCacheBatch.itemsCnt;
      }
      var maxBatchIndex = Math.floor(itemsCnt / this.options.batch_size) - 1;

      var minBatchIndex = 0;

      if (this.options.first) {
        maxBatchIndex = Math.round(maxBatchIndex * this.options.first / 100);
      }
      if (this.options.last) {
        minBatchIndex = Math.round(maxBatchIndex * (100 - this.options.last) / 100) + 1;
      }

      this.index = minBatchIndex;
      const iterator = {

        next: async () => {
          var targetCacheBatchIndex = Math.floor(this.index * this.options.batch_size / this.cache_batch_size);
          var nextCacheBatchIndex = Math.floor((this.index + 1) * this.options.batch_size / this.cache_batch_size);
          var indexWithinTheCacheBatch = ((this.index * this.options.batch_size) % this.cache_batch_size);
          var moreBatchesInTheCacheBatch = indexWithinTheCacheBatch + this.options.batch_size < this.cache_batch_size;
          var entierCacheBatch = lastLoadedCacheBatch;
          if (lastLoadedCacheBatchIndex == null || targetCacheBatchIndex != lastLoadedCacheBatchIndex)
            entierCacheBatch = JSON.parse(fs.readFileSync(`${cacheFileDir}${targetCacheBatchIndex}.json`));
          lastLoadedCacheBatch = entierCacheBatch;
          lastLoadedCacheBatchIndex = targetCacheBatchIndex;
          var itemsLeftInTheCacheBatchToAdd = Math.min(entierCacheBatch.xs.length - indexWithinTheCacheBatch, this.options.batch_size);
          var batch = { xs: entierCacheBatch.xs.slice(indexWithinTheCacheBatch, indexWithinTheCacheBatch + itemsLeftInTheCacheBatchToAdd), ys: entierCacheBatch.ys.slice(indexWithinTheCacheBatch, indexWithinTheCacheBatch + itemsLeftInTheCacheBatchToAdd) };
          if (batch.xs.length < this.options.batch_size && !moreBatchesInTheCacheBatch && fs.existsSync(`${cacheFileDir}${nextCacheBatchIndex}.json`)) {
            //since cache batch is bigger than a training batch
            //next cache batch should have enough items for the training batch
            targetCacheBatchIndex++;
            entierCacheBatch = JSON.parse(fs.readFileSync(`${cacheFileDir}${targetCacheBatchIndex}.json`));
            lastLoadedCacheBatch = entierCacheBatch;
            lastLoadedCacheBatchIndex = targetCacheBatchIndex;
            batch.xs = batch.xs.concat(entierCacheBatch.xs.slice(0, this.options.batch_size - batch.xs.length));
            batch.ys = batch.ys.concat(entierCacheBatch.ys.slice(0, this.options.batch_size - batch.ys.length));
          }

          this.index++;
          var done = this.index > maxBatchIndex || (!moreBatchesInTheCacheBatch && !fs.existsSync(`${cacheFileDir}${nextCacheBatchIndex}.json`));
          return { value: { xs: tf.tensor(batch.xs), ys: tf.tensor(batch.ys) }, done: done };
        }
      };

      return iterator;
    }
  }

}

async function trainModel(workerData) {
  try {
    var wguid = guidGenerator();
    console.log(`Worker ${wguid} started`);
    var phenotype = workerData.phenotype;
    const trainDataset =
      tf.data
        .generator(
          () => new DataSet(
            workerData.tensors.trainingDataSetSource.host,
            workerData.tensors.trainingDataSetSource.path,
            workerData.tensors.trainingDataSetSource.port, 
            workerData.tensors.trainingDataSetSource.cache_id, 
            workerData.tensors.trainingDataSetSource.cache_batch_size, 
            { first : (1 - workerData.validationSplit) * 100, batch_size: phenotype.batchSize }
          ).getNextBatchFunction()
        );
    const trainValidationDataset =
      tf.data
        .generator(
          () => new DataSet(
            workerData.tensors.trainingDataSetSource.host,
            workerData.tensors.trainingDataSetSource.path,
            workerData.tensors.trainingDataSetSource.port, 
            workerData.tensors.trainingDataSetSource.cache_id, 
            workerData.tensors.trainingDataSetSource.cache_batch_size, 
            { last : workerData.validationSplit * 100, batch_size: phenotype.batchSize }
          ).getNextBatchFunction()
        );        
    const valDataset =
      tf.data
        .generator(
          () => new DataSet(
            workerData.tensors.validationDataSetSource.host,
            workerData.tensors.validationDataSetSource.path,
            workerData.tensors.validationDataSetSource.port,
            workerData.tensors.validationDataSetSource.cache_id, 
            workerData.tensors.validationDataSetSource.cache_batch_size, 
            { batch_size: phenotype.batchSize }
          ).getNextBatchFunction()
        );

    var modelAbortThreshold = workerData.modelAbortThreshold;
    var modelTrainingTimeThreshold = workerData.modelTrainingTimeThreshold;
    var validationSplit = workerData.validationSplit;

    var lossThresholdAbortCnt = 0;
    var model = null;
    do {
      const modelData = JSON.parse(workerData.modelJson);
      const weightData = new Uint8Array(Buffer.from(modelData.weightData, "base64")).buffer;
      model = await tf.loadLayersModel(tf.io.fromMemory({
        modelTopology: modelData.modelTopology,
        weightSpecs: modelData.weightSpecs,
        weightData: weightData
      }));

      var trainLogs = [];
      var lossThresholdAbort = false;
      var errorAbort = false;
      var epochTimeStart;

      model.compile(
        { optimizer: optimizerBuilderFunction(phenotype.optimizer, phenotype.learningRate), loss: phenotype.loss });
      var trainingStartTime = Date.now();
      await model.fitDataset(
        trainDataset,
        //tensors.trainFeatures, tensors.trainTarget,
        {
          verbose: false,
          //batchSize: phenotype.batchSize,
          epochs: phenotype.epochs,
          //validationSplit: validationSplit,
          validationData: trainValidationDataset,
          callbacks: {
            onEpochBegin: (epoch, logs) => {
              epochTimeStart = Date.now();
            },
            onEpochEnd: async (epoch, logs) => {
              trainLogs.push(logs);
              if (isNaN(logs.val_loss)) {
                console.log(`Early model loss is NaN abort. Epoch ${epoch} `);
                errorAbort = true;
                throw Error("Loss is NaN");
              }
              if (modelTrainingTimeThreshold && Date.now() > trainingStartTime + modelTrainingTimeThreshold * 1000) {
                console.log(`Early model training timeout abort. Epoch ${epoch} `);
                errorAbort = true;
                throw Error("Model training timeout abort");
              }
              if (modelTrainingTimeThreshold && epochTimeStart && (phenotype.epochs - 1 > epoch)
                && ((Date.now() - epochTimeStart) * (phenotype.epochs - epoch - 1)) > modelTrainingTimeThreshold * 1000) {
                console.log(`Early model training timeout abort based on prior epoch time. Epoch ${epoch} `);
                errorAbort = true;
                throw Error("Model training timeout abort");
              }  
              if (modelAbortThreshold && trainLogs.length > modelAbortThreshold && trainLogs[trainLogs.length - modelAbortThreshold].val_loss <= logs.val_loss) {
                //console.log(`Early model training abort(${lossThresholdAbortCnt+1}). Epoch ${epoch}. loss compare ` + trainLogs[trainLogs.length - modelAbortThreshold].val_loss + " <= " + logs.val_loss);
                lossThresholdAbort = true;
                throw Error("Early training abort");
              }
            }
          }
        }).catch((err) => {
          if (!lossThresholdAbort && !errorAbort) console.log(err)
        });

      if (errorAbort) {
        return { validationLoss: NaN, phenotype: phenotype };
      }
      if (lossThresholdAbort) {
        phenotype.epoch = trainLogs.length - modelAbortThreshold + 1;

        lossThresholdAbortCnt++;
        if (lossThresholdAbortCnt > 1) {
          console.log(`Early model training abort`);
          return { validationLoss: NaN, phenotype: phenotype };
        }
      }
    } while (lossThresholdAbort)


    const result = await model.evaluateDataset(
      valDataset
      //tensors.testFeatures, tensors.testTarget, { batchSize: phenotype.batchSize }
    );
    const testLoss = parseFloat(result.dataSync()[0].toFixed(4));
    const trainLoss = trainLogs[trainLogs.length - 1].loss.toFixed(4);
    const valLoss = trainLogs[trainLogs.length - 1].val_loss.toFixed(4);

    phenotype.validationLoss = testLoss;

    console.log(`Model training completed ${phenotype._id} . post training eval loss ${testLoss}, training validation-set loss: ${valLoss}, train-set loss: ${trainLoss}`);
    console.log("============================================================");

    let resultingModelData = await model.save(tf.io.withSaveHandler(async modelArtifacts => modelArtifacts));
    resultingModelData.weightData = Buffer.from(resultingModelData.weightData).toString("base64");
    const jsonStr = JSON.stringify(resultingModelData);

    return { validationLoss: testLoss, phenotype: phenotype, modelJson: jsonStr };
  }
  catch (err) {
    console.log(`Error: ${err} stack: ${err.stack}`)
    return { validationLoss: NaN, phenotype: phenotype };
  }
}

async function readQueue() {
  console.log(`${Date.now()} readQueue() start`);
  return new Promise(async function (resolve, reject) {
    try {
      const connection = await amqp.connect(queueUrl);
      const channel = await connection.createChannel();

      process.once("SIGINT", async () => {
        await channel.close();
        await connection.close();
        resolve(null);
      });

      await channel.assertQueue(inputQueue, { durable: false });
      await channel.prefetch(1);

      await channel.consume(
        inputQueue,
        async (message) => {
          if (message) {
            resolve(JSON.parse(message.content.toString()));
          }
          else {
            resolve(null);
          }
          await channel.ack(message);
          await channel.close();
          await connection.close();
          console.log(`${Date.now()} readQueue() closing channel`);
        }
      );

    } catch (err) {
      reject(err);
    }
  });
}


async function writeQueue(id, tfjsJobResponse) {
  return new Promise(async function (resolve, reject) {
    var outputQueue = `${outputQueuePrefix}-${id}`;
    console.log(`${Date.now()} writeQueue ${outputQueue}`);
    let connection;
    try {
      connection = await amqp.connect(queueUrl);
      const channel = await connection.createChannel();

      await channel.assertQueue(outputQueue, { durable: false });
      channel.sendToQueue(outputQueue, Buffer.from(JSON.stringify(tfjsJobResponse)));
      await channel.close();
      resolve();
    } catch (err) {
      console.warn(err);
      reject(err);
    } finally {
      if (connection) await connection.close();
    }
  });
}

async function main() {
  console.log(`${Date.now()} main() start`);
  while (true) {
    try {
      var tfjsJob = await readQueue();
      var response = await trainModel(tfjsJob.workerData);
      await writeQueue(response.phenotype._id, response);
    }
    catch (err) {
      console.log(`Worker trainModel Error: ${err} stack: ${err.stack}`)
    }
  }
}

main();

File: distributed-training/kubernetes/docker/python_worker/worker.py
import os
import uuid
import tensorflow as tf
from pathlib import Path
import urllib.request
import json
import math
from datetime import datetime
from rabbitmq import RabbitMQ
import traceback
from utils import DictToObj
import time


SETTINGS = { "TEST_MODE" : False}


CACHE_STORAGE = "./_runtime/cache/"
MODEL_STORAGE = "shared/models/"

#amqp://rabbitmq-service:5672
RABBITMQ_HOST = None #os.getenv('RABBITMQ_HOST', 'rabbitmq-service')
RABBITMQ_PORT = None #int(os.getenv('RABBITMQ_PORT', 5672))

inputQueue = None #os.environ["JOB_NAME"] + "-INPUT"
outputQueuePrefix = None #os.environ["JOB_NAME"] + "-OUTPUT"

def guidGenerator():
    return str(uuid.uuid4())                                                  

class DataSet:
    def __init__(self, host, path, port, cache_id, cache_batch_size, options):
            self.host = host
            self.path = path
            self.port = port
            self.cache_id = cache_id
            self.cache_batch_size = cache_batch_size
            self.index = 0
            defaultOptions = { 'first': 0, 'last': 0, 'batch_size': 32 }
            self.options = defaultOptions | options
            print(self.options)
            if (int(self.cache_batch_size) <= int(self.options["batch_size"])):
                raise Exception("cache_batch_size should always be bigger than options.batch_size")
            
            self.cacheFileDir = f"{CACHE_STORAGE}{self.cache_id}/"
            self.lastLoadedCacheBatch = None
            self.lastLoadedCacheBatchIndex = None
            self.shape = None
            
    
    def download(self):
         itemsCnt=0
         if(not os.path.isdir(self.cacheFileDir)):
              tmpFolderGuid = guidGenerator()
              self.cacheFileDir = f"{CACHE_STORAGE}{tmpFolderGuid}_{self.cache_id}/"
              path = Path(self.cacheFileDir)
              path.mkdir(parents=True, exist_ok=True)
              batch_index = 0
              source_path = f"{self.path}?index={batch_index}&cache_batch_size={self.cache_batch_size}"
              itemsCnt = 0
              res = { "done": False }
              while(not res["done"]):
                   print(f"http://{self.host}:{self.port}{source_path}")
                   res = json.loads(urllib.request.urlopen(f"http://{self.host}:{self.port}{source_path}").read())
                   with open(f"{self.cacheFileDir}{batch_index}.json", "w") as f:
                        f.write(res["value"])
                   itemsCnt += len(json.loads(res["value"])["xs"])
                   batch_index=batch_index+1
                   source_path = f"{self.path}?index={batch_index}&cache_batch_size={self.cache_batch_size}"
                   if res["done"]:
                        with open(f"{self.cacheFileDir}0.json", "r+") as f:
                            firstBatch = json.loads(f.read())
                            firstBatch["itemsCnt"] = itemsCnt
                            f.seek(0)
                            f.write(json.dumps(firstBatch))
              os.rename(self.cacheFileDir, f"{CACHE_STORAGE}{self.cache_id}/")
              self.cacheFileDir = f"{CACHE_STORAGE}{self.cache_id}/"
         if (itemsCnt==0):
            with open(f"{self.cacheFileDir}0.json", "r") as f:
                self.entierCacheBatch = json.loads(f.read())
                itemsCnt = self.entierCacheBatch["itemsCnt"]
                
         self.maxBatchIndex = math.floor(itemsCnt / self.options["batch_size"]) - 1

         self.minBatchIndex = 0

         if (self.options["first"]):
              if SETTINGS["TEST_MODE"]:     
                self.options["first"] = 0.1

              self.maxBatchIndex = round(self.maxBatchIndex * self.options["first"] / 100)
                
         if (self.options["last"]):
              if SETTINGS["TEST_MODE"]:    
                self.options["last"] = 0.1
              
              self.minBatchIndex = round(self.maxBatchIndex * (100 - self.options["last"]) / 100) + 1    

         print(f"self.minBatchIndex: {self.minBatchIndex}, self.maxBatchIndex: {self.maxBatchIndex}, itemsCnt: {itemsCnt}")    

         self.__next__()
         self.index = 0

    def __iter__(self):
         self.index = 0
         return self

    def __next__(self):
          targetCacheBatchIndex = math.floor(self.index * self.options["batch_size"] / self.cache_batch_size)
          nextCacheBatchIndex = math.floor((self.index + 1) * self.options["batch_size"] / self.cache_batch_size)
          indexWithinTheCacheBatch = ((self.index * self.options["batch_size"]) % self.cache_batch_size)
          moreBatchesInTheCacheBatch = indexWithinTheCacheBatch + self.options["batch_size"] < self.cache_batch_size
          
          self.entierCacheBatch = self.lastLoadedCacheBatch
          if (self.lastLoadedCacheBatchIndex == None or targetCacheBatchIndex != self.lastLoadedCacheBatchIndex):
            with open(f"{self.cacheFileDir}{targetCacheBatchIndex}.json", "r") as f:
                self.entierCacheBatch = json.loads(f.read())
          self.lastLoadedCacheBatch = self.entierCacheBatch
          self.lastLoadedCacheBatchIndex = targetCacheBatchIndex
          itemsLeftInTheCacheBatchToAdd = min(len(self.entierCacheBatch["xs"]) - indexWithinTheCacheBatch, self.options["batch_size"])
          batch = { "xs": self.entierCacheBatch["xs"][slice(indexWithinTheCacheBatch, indexWithinTheCacheBatch + itemsLeftInTheCacheBatchToAdd)], "ys": self.entierCacheBatch["ys"][slice(indexWithinTheCacheBatch, indexWithinTheCacheBatch + itemsLeftInTheCacheBatchToAdd)] }
          file_path = Path(f"{self.cacheFileDir}{nextCacheBatchIndex}.json")
          if (len(batch["xs"]) < self.options["batch_size"] and not moreBatchesInTheCacheBatch and file_path.exists()):
                #since cache batch is bigger than a training batch
                #next cache batch should have enough items for the training batch
                targetCacheBatchIndex=targetCacheBatchIndex+1
                with open(f"{self.cacheFileDir}{targetCacheBatchIndex}.json", "r") as f:
                    entierCacheBatch = json.loads(f.read())
                self.lastLoadedCacheBatch = entierCacheBatch
                self.lastLoadedCacheBatchIndex = targetCacheBatchIndex
                batch["xs"] = batch["xs"] + self.entierCacheBatch["xs"][slice(0, self.options["batch_size"] - len(batch["xs"]))]
                batch["ys"] = batch["ys"] + self.entierCacheBatch["ys"][slice(0, self.options["batch_size"] - len(batch["ys"]))]
          

          self.index=self.index+1
          file_path = Path(f"{self.cacheFileDir}{nextCacheBatchIndex}.json")
          done = self.index > self.maxBatchIndex or (not moreBatchesInTheCacheBatch and not file_path.exists())
          #print(f"self.index: {self.index}, done: {done}, self.maxBatchIndex: {self.maxBatchIndex}, targetCacheBatchIndex: {targetCacheBatchIndex}, nextCacheBatchIndex: {nextCacheBatchIndex}, indexWithinTheCacheBatch: {indexWithinTheCacheBatch}, moreBatchesInTheCacheBatch: {moreBatchesInTheCacheBatch}")
          if self.shape == None:
               vshape = tf.constant(batch["xs"]).shape.as_list()
               vshape[0] = None
               self.input_shape = tf.TensorShape(vshape)
               vshape = tf.constant(batch["ys"]).shape.as_list()
               vshape[0] = None
               self.output_shape = tf.TensorShape(vshape)               
            #    print(f"input shape: {self.input_shape}")
            #    print(f"output shape: {self.output_shape}")
          if done: 
               raise StopIteration
          #return { "value": { "xs": tf.constant(batch["xs"]), "ys": tf.constant(batch["ys"]) }, "done": done }
          return batch["xs"],batch["ys"]


def trainModel(wguid, model, workerdata, trainDS, trainDS_genRes, trainValidationDS, trainValidationDS_genRes, validationDS, validationDS_genRes, patience=2):
    early_stopping = tf.keras.callbacks.EarlyStopping(monitor='val_loss',
                                                        patience=patience,
                                                        mode='min')  
    history = model.fit(trainDS, epochs=workerdata["phenotype"]["epochs"], steps_per_epoch = trainDS_genRes.maxBatchIndex - trainDS_genRes.minBatchIndex, validation_steps = trainValidationDS_genRes.maxBatchIndex - trainValidationDS_genRes.minBatchIndex,
                        validation_data=trainValidationDS,
                        callbacks=[early_stopping])  
    
    val_performance = model.evaluate(validationDS, return_dict=True)
    loss = val_performance["loss"]
    workerdata["phenotype"]["validationLoss"] = loss
    keras_model_path = f'{CACHE_STORAGE}{wguid}.keras'
    model.save(keras_model_path)
    jsonStr = None
    with open(keras_model_path, 'rb') as file:
        jsonStr = file.read().hex()  

    res = { "validationLoss": loss, "phenotype": workerdata["phenotype"], "modelJson": jsonStr }

    return res

buildModel = None

def readQueue():
    try:
        def callback(ch, method, properties, body):
            global buildModel
            if body != None:
                try:
                    print(body)
                    wguid = guidGenerator()
                    print(f"Worker {wguid} started")
                    message = json.loads(body.decode("utf-8"))
                    workerData = message["workerData"]
                    trainDS_genRes = ds_gen(workerData["tensors"]["trainingDataSetSource"], { "batch_size": workerData["phenotype"]["batchSize"], "first": (1-workerData["validationSplit"]) * 100 })
                    trainValidationDS_genRes = ds_gen(workerData["tensors"]["trainingDataSetSource"], { "batch_size": workerData["phenotype"]["batchSize"], "last": workerData["validationSplit"] * 100 })
                    validationDS_genRes = ds_gen(workerData["tensors"]["validationDataSetSource"], { "batch_size": workerData["phenotype"]["batchSize"] })

                    if buildModel == None:
                        local_ns = {}
                        exec(workerData["modelJson"]["buildModel"], None, local_ns)
                        buildModel = local_ns["buildModel"]
                    model = buildModel(workerData, trainDS_genRes.input_shape)
                    trainDS_genRes = ds_gen(workerData["tensors"]["trainingDataSetSource"], { "batch_size": workerData["phenotype"]["batchSize"], "first": (1-workerData["validationSplit"]) * 100 })
                    trainValidationDS_genRes = ds_gen(workerData["tensors"]["trainingDataSetSource"], { "batch_size": workerData["phenotype"]["batchSize"], "last": workerData["validationSplit"] * 100 })
                    validationDS_genRes = ds_gen(workerData["tensors"]["validationDataSetSource"], { "batch_size": workerData["phenotype"]["batchSize"] })

                    trainDS = tf.data.Dataset.from_generator(generator=lambda:ds_gen(workerData["tensors"]["trainingDataSetSource"], { "batch_size": workerData["phenotype"]["batchSize"], "first": (1-workerData["validationSplit"]) * 100 }),
                        output_types=(tf.float32, tf.float32),
                        output_shapes = (trainDS_genRes.input_shape,trainDS_genRes.output_shape))
                    trainValidationDS = tf.data.Dataset.from_generator(generator=lambda:ds_gen(workerData["tensors"]["trainingDataSetSource"], { "batch_size": workerData["phenotype"]["batchSize"], "last": workerData["validationSplit"] * 100 }),
                        output_types=(tf.float32, tf.float32),
                        output_shapes = (trainValidationDS_genRes.input_shape,trainValidationDS_genRes.output_shape))  
                    validationDS = tf.data.Dataset.from_generator(generator=lambda:ds_gen(workerData["tensors"]["validationDataSetSource"], { "batch_size": workerData["phenotype"]["batchSize"] }),
                        output_types=(tf.float32, tf.float32),
                        output_shapes = (validationDS_genRes.input_shape,validationDS_genRes.output_shape))   
                    trainRes = trainModel(wguid, model, workerData, trainDS, trainDS_genRes, trainValidationDS, trainValidationDS_genRes, validationDS, validationDS_genRes)                 
                    
                    writeQueue(trainRes["phenotype"]["_id"], trainRes)
                except Exception as callback_err:
                    print(traceback.format_exc())
            time.sleep(3)
            #readQueue()

        rabbitmq = RabbitMQ(RABBITMQ_HOST, RABBITMQ_PORT)
        rabbitmq.consume(inputQueue, callback)
        #callback(None, None, None, rabbitmq.get(inputQueue))
    except Exception as err:
        print(traceback.format_exc())
        time.sleep(3)
        print("trying to re-establish connection")
        readQueue()

def writeQueue(id, tfjsJobResponse):
    outputQueue = f"{outputQueuePrefix}-{id}"
    rabbitmq = RabbitMQ(RABBITMQ_HOST, RABBITMQ_PORT)
    rabbitmq.publish(outputQueue, json.dumps(tfjsJobResponse))
    rabbitmq.close()  

def ds_gen(data_source, options):
     trainDS = DataSet(data_source["host"], data_source["path"], data_source["port"], data_source["pre_cache"], data_source["cache_batch_size"], options)
     trainDS.download()
     return trainDS

def optimizerBuilderFunction(optimizer, learningRate):
    print(f'optimizerBuilderFunction({optimizer}, {learningRate})')
    match optimizer:
         case "sgd":
              return tf.keras.optimizers.SGD(learning_rate=learningRate)
         case "adagrad":
              return tf.keras.optimizers.Adagrad(learning_rate=learningRate)
         case "adadelta":
              return tf.keras.optimizers.Adadelta(learning_rate=learningRate)
         case "adam":
              return tf.keras.optimizers.Adam(learning_rate=learningRate)
         case "adamax":
              return tf.keras.optimizers.Adamax(learning_rate=learningRate)
         case "rmsprop":
              return tf.keras.optimizers.RMSprop(learning_rate=learningRate)  

def initializers(initializer):
    match initializer:
        case 'leCunNormal':
            return tf.keras.initializers.LecunNormal()
        case 'glorotNormal':
            return tf.keras.initializers.GlorotNormal()
        case 'glorotUniform':
            return tf.keras.initializers.GlorotUniform()
        case 'heNormal':
            return tf.keras.initializers.HeNormal()
        case 'heUniform':
            return tf.keras.initializers.HeUniform()
        case 'leCunUniform':
            return tf.keras.initializers.LecunUniform()
        case 'randomNormal':
            return tf.keras.initializers.RandomNormal()       
        case 'randomUniform':
            return tf.keras.initializers.RandomUniform()    
        case 'truncatedNormal':
            return tf.keras.initializers.TruncatedNormal()    
        case 'varianceScaling':
            return tf.keras.initializers.VarianceScaling()                                                                    

def lossBuilderFunction(loss):
     match loss:
          case "meanAbsoluteError":
               return tf.keras.losses.MeanAbsoluteError()
          case "meanSquaredError":
               return tf.keras.losses.MeanSquaredError()

def main():  
    global SETTINGS, RABBITMQ_HOST, RABBITMQ_PORT, inputQueue, outputQueuePrefix
    print(f"main() -> SETTINGS: {SETTINGS}")

    if SETTINGS["TEST_MODE"]:
        RABBITMQ_HOST = os.getenv('RABBITMQ_HOST', '127.0.0.1')
        RABBITMQ_PORT = int(os.getenv('RABBITMQ_PORT', 30000))        
    else:
        RABBITMQ_HOST = os.getenv('RABBITMQ_HOST', 'rabbitmq-service')
        RABBITMQ_PORT = int(os.getenv('RABBITMQ_PORT', 5672))
    inputQueue = os.environ["JOB_NAME"] + "-INPUT"
    outputQueuePrefix = os.environ["JOB_NAME"] + "-OUTPUT"
    print(f"inputQueue: {inputQueue}")
    print(f"outputQueuePrefix: {outputQueuePrefix}")
    readQueue()